src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt

```diff
@@ -0,0 +1,369 @@
+        val filterParams = mutableMapOf<String, Any?>()
+
+        for (argument in field.arguments) {
+            if (argument.name == ProjectionBase.FILTER) {
+                val parsedQuery = ParsedQuery(argument.value as ObjectValue, type)
+                withWithoutWhere = NestingLevelHandler(parsedQuery, false, rootNode, variable, { readingWithoutWhere },
+                        type, argument.value, filterParams, linkedSetOf(rootNode.requiredSymbolicName))
+                    .parseFilter()
+            } else {
+                throw OptimizedQueryException()
+            }
+        }
+        return withWithoutWhere to filterParams
+    }
+
+    /**
+     * @param parsedQuery the internal representation of the parsed query for this nesting level
+     * @param useDistinct should the current node be distinct (if true: renders WITH DISTINCT currentNode)
+     * @param current the current node
+     * @param variablePrefix the prefix to prepend to new variables
+     * @param type the type of <code>current</code>
+     * @param value the value passed to the graphQL field
+     * @param filterParams the map to store required filter params into
+     * @param parentPassThroughWiths all the nodes, required to be passed through via WITH
+     */
+    class NestingLevelHandler(
+            private val parsedQuery: ParsedQuery,
+            private val useDistinct: Boolean,
+            private val current: PropertyContainer<*>,
+            private val variablePrefix: String,
+            private val readingWithoutWhereFactory: () -> StatementBuilder.OngoingReading,
+            private val type: GraphQLFieldsContainer,
+            private val value: Value<*>?,
+            private val filterParams: MutableMap<String, Any?>,
+            private val parentPassThroughWiths: Collection<Expression>
+    ) {
+
+        /**
+         * @param additionalFilter additional filters to be applied to the where
+         */
+        fun parseFilter(additionalFilter: ((StatementBuilder.ExposesWith) -> StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere)? = null): StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere {
+            if (value !is ObjectValue) {
+                throw OptimizedQueryException()
+            }
+            // WHERE MATCH all predicates for current
+            // WITH x
+            var query = addWhere(additionalFilter)
+
+            // Handle all quantifier (ALL / NONE / SINGLE / ANY)
+            query = handleRelation(query)
+
+            // Handle OR / AND
+            return handleCombinations(query)
+        }
+
+        private fun addWhere(
+                additionalFilter: ((StatementBuilder.ExposesWith) -> StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere)? = null
+        ): StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere {
+            val reading = readingWithoutWhereFactory()
```

Comment:
reading is an odd name,  can we find something better that resonates with what it is?